doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+end, you will need to re-run mkcloud after you have created your cloud:
+
+```
+nodenumberlonelynode=1 mkcloud setuplonelynodes crowbar_register
+
+```
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
+You will find the currently running Crowbar code (both Rails application
+components and chef recipes) in the `/opt/dell` directory. We will provide
+detailed locations for all important components of a Barclamp in the _Basic
+Implementation_ section below.
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
+
+### Preparations for Crowbar Node
+
+These are recommendations, but we'll assume you followed them for the remainder
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by the ID of the oldest
+# commit on your branch. Running
+#
+#   git log --format='%H %s' $branch..HEAD
+#
+# where $branch is the upstream branch you are developing against such as
+# `master` or `stable/3.0`, should give you a list of original commits. This
+# assumes $branch has not been improperly cherry-picked into or rebased in your
+# fork, though. It also assumes that you are not working on your fork's
+# `master` branch (you shouldn't, but people have been known to do this
+# anyway). Since all of these may occasionally happen, this guide leaves it up
+# to the user to set $COMMIT as they see fit rather than trying to determine
+# its value automatically through the command mentioned above.
+
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert your own barclamp's name here:
+export BARCLAMP_NAME=mybarclamp
+
+# Insert a space delimited list of the Barclamps you are working on here (only
+# required if your barclamp requires changes to other barclamps as well).
+
+export BARCLAMPS="$BARCLAMP_NAME"
+
+# Insert the path to your barclamp collection checkout here (in case it's
+# somewhere other than /root/crowbar-openstack). This directory must be a git
+# working copy of the crowbar-openstack fork and branch you are working on.
+
+export BARCLAMP_FORK_PATH=/root/crowbar-openstack
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
```

Comment:
Again, this would not be necessary if developing from your native dev environment, but if you are going to do this, it's better to do it via `git config`, so that it applies for all git commands, not just when you are running this script or running git commands in a shell where you forgot to first run these `export` commands.